src/lib.rs

```diff
@@ -566,60 +574,90 @@ impl Deref for SerializedSignature {
 
 impl Eq for SerializedSignature {}
 
-impl<C> Eq for Secp256k1<C> { }
+impl<'buf, C> Eq for Secp256k1<'buf, C> { }
 
-impl<C> Drop for Secp256k1<C> {
+impl<'buf, C> Drop for Secp256k1<'buf, C> {
     fn drop(&mut self) {
-        unsafe { ffi::secp256k1_context_destroy(self.ctx); }
+        match self._buf {
+            None => unsafe { ffi::secp256k1_context_destroy(self.ctx) },
+            Some(_) => unsafe { ffi::secp256k1_context_preallocated_destroy(self.ctx) },
+        }
     }
 }
 
-impl fmt::Debug for Secp256k1<SignOnly> {
+impl<'buf> fmt::Debug for Secp256k1<'buf, SignOnly> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "<secp256k1 context {:?}, signing only>", self.ctx)
     }
 }
 
-impl fmt::Debug for Secp256k1<VerifyOnly> {
+impl<'buf> fmt::Debug for Secp256k1<'buf, VerifyOnly> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "<secp256k1 context {:?}, verification only>", self.ctx)
     }
 }
 
-impl fmt::Debug for Secp256k1<All> {
+impl<'buf> fmt::Debug for Secp256k1<'buf, All> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "<secp256k1 context {:?}, all capabilities>", self.ctx)
     }
 }
 
-impl Secp256k1<All> {
+impl<'buf> Secp256k1<'buf, All> {
+    const FLAGS: c_uint = ffi::SECP256K1_START_SIGN | ffi::SECP256K1_START_VERIFY;
     /// Creates a new Secp256k1 context with all capabilities
-    pub fn new() -> Secp256k1<All> {
-        Secp256k1 { ctx: unsafe { ffi::secp256k1_context_create(ffi::SECP256K1_START_SIGN | ffi::SECP256K1_START_VERIFY) }, phantom: PhantomData }
+    pub fn new() -> Secp256k1<'buf, All> {
+        Secp256k1 { ctx: unsafe { ffi::secp256k1_context_create(Self::FLAGS) }, phantom: PhantomData, _buf: None }
```

Comment:
As discussed on IRC, I'd like to allocate our own memory in Rust and then call the prealloc functions.